fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/search/LuceneOptimizedMultiFieldQueryParser.java

```diff
@@ -72,25 +86,145 @@ protected Query getFieldQuery(String field, String queryText, int slop) throws P
+            return IntPoint.newExactQuery(field, point.intValue());
+        } else if (Long.class.equals(cfg.getType())) {
+            return LongPoint.newExactQuery(field, point.longValue());
+        } else if (Double.class.equals(cfg.getType())) {
+            return DoublePoint.newExactQuery(field, point.doubleValue());
+        } else if (Float.class.equals(cfg.getType())) {
+            return FloatPoint.newExactQuery(field, point.floatValue());
+        } else {
+            throw new ParseException("Unknown numeric type: " + cfg.getType().getCanonicalName());
+        }
+    }
+
+
+    @Override
+    @SuppressWarnings("PMD.PreserveStackTrace") //it isn't possible with Lucene's exception API
+    protected Query getRangeQuery(final String field,
+                                  final String part1,
+                                  final String part2,
+                                  final boolean startInclusive,
+                                  final boolean endInclusive) throws ParseException {
+        /*
+         * Lucene doesn't really understand types, so unless we tell it that we are looking at numeric-valued
+         * data points in our scan, it will parse everything as text, which results in incorrect results
+         * being returned over range scans.
+         *
+         * To avoid this, we use a PointConfig map (an idea taken from Lucene's StandardAnalyzer). This allows
+         * us to specify the data types for individual fields in Lucene, which we use here to create the correct
+         * type of Query object for range scans.
+         */
+        PointsConfig cfg = pointsConfig.get(field);
+        if (cfg == null) {
+            return super.getRangeQuery(field, part1, part2, startInclusive, endInclusive);
+        } else {
+            NumberFormat format = cfg.getNumberFormat();
+            Number start;
+            Number end;
+            try {
+                start = format.parse(part1);
+                end = format.parse(part2);
+            } catch (java.text.ParseException pe) {
+                throw new ParseException(QueryParserMessages.COULD_NOT_PARSE_NUMBER);
+            }
+
+            if (Integer.class.equals(cfg.getType())) {
+                int s = start.intValue();
+                int e = end.intValue();
+                //lucene range queries are inclusive, so adjust ranges as needed
+                if (!startInclusive) {
+                    s = Math.addExact(s, 1);
+                }
+                if (!endInclusive) {
+                    e = Math.addExact(e, -1);
+                }
+
+                return IntPoint.newRangeQuery(field, s, e);
+            } else if (Long.class.equals(cfg.getType())) {
+                long s = start.longValue();
+                long e = end.longValue();
+                //lucene range queries are inclusive, so adjust ranges as needed
+                if (!startInclusive) {
+                    s = Math.addExact(s, 1);
```

Comment:
nit: addExact will throw exception if add results in a number > MAX_INT but the type is long here...

Reply:
nit: sanity checks silly here? e.g. assert e > s and e > 0?

Reply:
There's an `addExact(long, long)`, too, going to `Long.MAX_VALUE`.

Reply:
Missed that @MMcM (Ignore my first comment @scottfines )

Reply:
put a Parse check in for each type